ethers-contract/src/contract.rs

```diff
@@ -145,37 +145,73 @@ use ethers_core::types::TransactionRequest;
 ///
 /// _Disclaimer: these above docs have been adapted from the corresponding [ethers.js page](https://docs.ethers.io/ethers.js/html/api-contract.html)_
 ///
+/// # Usage Note
+///
+/// `ContractInternal` accepts any client that implements `B: Borrow<M>` where
+/// `M :Middleware`. Previous `Contract` versions used only arcs, and relied
+/// heavily on [`Arc`]. Due to constraints on the [`ContractCallInternal`] type,
+/// calling contracts requires a `B: Borrow<M> + Clone`. This is fine for most
+/// middlware. However, when `B` is an owned middleware that is not Clone, we
+/// cannot issue contract calls. Some notable exceptions:
+///
+/// - `NonceManagerMiddleware`
+/// - `SignerMiddleware` (when using a non-Clone Signer)
+///
+/// When using non-Clone middlewares, instead of instantiating a contract that
+/// OWNS the middlware, pass the contract a REFERENCE to the middleware. This
+/// will fix the trait bounds issue (as `&M` is always `Clone`).
+///
+/// We expect to fix this fully in a future version
+///
 /// [`abigen`]: macro.abigen.html
 /// [`Abigen` builder]: struct.Abigen.html
 /// [`event`]: method@crate::Contract::event
 /// [`method`]: method@crate::Contract::method
 #[derive(Debug)]
-pub struct Contract<M> {
+pub struct ContractInternal<B, M> {
```

Comment:
The same concerns re naming here.
`Internal` is likely confusing, but for this one, I don't have any suggestions at this point.

hmm, compared to `BaseContract` this type represents an `Existing`, `Deployed` contract since it has an `address`, so perhaps we include this in the name?

Reply:
Naming for `Contract` is inherited from ethers.js, and implies an interface with an address and a provider. I believe `BaseContract` was backfilled to make sense given `Contract` has a connection to the chain.

proposed better nomenclature:

- `Interface` -- currently called `BaseContract` provides encoding/decoding functionality
   - alias to `BaseContract` for backwards compat
   - maybe de-duplicate the hashmap via a global cache so we stop doing a bunch of allocs on clone?
- `ContractInstance` -- currently `ContractInternal`. Wraps `BaseContract` with an address and a connection
  - alias version with arc to `Contract` for backwards compat